File: barthe/exposition/item.php
<?php 
namespace Barthe\Exposition;

require_once('config.php');
require_once('album.php');
require_once('photo.php');

abstract class Item 
{
	// Private members
	protected $path;
	protected $title;

	// Constructor
	protected function __construct($path, $title=NULL) 
	{		
		$this->path = $path;
		$this->title = $title;
		if (! $this->title)
			$this->title = basename($this->path);
	}

	// Factory method
	// ### TODO: Add option to stop recursion
	static function createItem($path) 
	{
		// Create album or photo depending on filesystem
		if (is_dir($path)) {
			return new Album($path);
		} else if (file_exists($path)) {
			$ext = strtolower(pathinfo($path, PATHINFO_EXTENSION));
			if (in_array($ext, Config::PHOTO_EXTENSIONS())) {
				return new Photo($path);
			}
		}

		// Nothing was created
		return NULL;
	}

	// Accessors
	public function getPath() 
	{
		return $this->path;
	}

	public function getTitle() 
	{
		return $this->title;
	}
}

?>

File: src/Exposition/Models/Item.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Exposition.Models
{
    abstract public class Item
    {
        public const string TYPE_ALBUM = "album";
        public const string TYPE_PHOTO = "photo";

        public string Title { get; }
        public string Path { get; }
        public abstract string Type { get; }

        public Item(string path, string title)
        {
            Contract.RequireNotNull(title);
            Contract.RequireNotNull(path);
            this.Title = title;
            this.Path = path;
        }
    }
}
